### PR TITLE
Openldap to 389ds migration tools test case

### PR DESCRIPTION
**Pull request by coolgw**
This test case include basic check on openldap to 389 migration tools.

- Related ticket: https://progress.opensuse.org/issues/81352
- Needles: na
- Verification run: https://openqa.suse.de/tests/5607399#step/openldap_to_389ds/72
New run after update, for failed part will submit an issue:
https://openqa.suse.de/tests/5620935#step/openldap_to_389ds/112

### CODE REVIEW

File: schedule/migration/openldap_to_389ds.yaml
Comment (b10n1k) [inline schedule/migration/openldap_to_389ds.yaml:3]: extra whitespace

File: tests/migration/openldap_to_389ds.pm
Comment (b10n1k) [inline tests/migration/openldap_to_389ds.pm:30]: $self->select_serial_terminal instead? wdyt?

File: tests/migration/openldap_to_389ds.pm
Comment (b10n1k) [inline tests/migration/openldap_to_389ds.pm:37]: you install sssd_tools again here. is this intentional?

File: data/openldap_to_389ds/instance.inf
Comment (rfan1) [inline data/openldap_to_389ds/instance.inf:5]: Wei,
I suggest you use your own password here

File: data/openldap_to_389ds/slapd.conf
Comment (rfan1) [inline data/openldap_to_389ds/slapd.conf:38]: You didn't test TLS right, is there any bugs or by plan?

File: data/openldap_to_389ds/sssd.conf
Comment (rfan1) [inline data/openldap_to_389ds/sssd.conf:19]: Is this ca file used during your test? 
I noticed that ldap_uri = ldap (rather than ldaps)

File: tests/migration/openldap_to_389ds.pm
Comment (rfan1) [inline tests/migration/openldap_to_389ds.pm:34]: I am wondering that you wanted to run rpm -qi?

File: tests/migration/openldap_to_389ds.pm
Comment (rfan1) [inline tests/migration/openldap_to_389ds.pm:60]: Why ignore_failure here?

File: tests/migration/openldap_to_389ds.pm
Comment (rfan1) [inline tests/migration/openldap_to_389ds.pm:82]: Is any reason add wait time here?

File: data/openldap_to_389ds/instance.inf
Comment (foursixnine) [reply data/openldap_to_389ds/instance.inf:5]: I would suggest to replace this pasword (using sed or something before the migration) with the root password defined in the testapi. But it's just a minor suggestion in this case.

File: tests/migration/openldap_to_389ds.pm
Comment (foursixnine) [inline tests/migration/openldap_to_389ds.pm:36]: please remember to put this in a branch, sle. 

```suggestion
    add_suseconnect_product('sle-module-legacy') if is_sle;
```

File: tests/migration/openldap_to_389ds.pm
Comment (foursixnine) [inline tests/migration/openldap_to_389ds.pm:61]: Same as above

File: schedule/migration/openldap_to_389ds.yaml
Comment (coolgw) [reply schedule/migration/openldap_to_389ds.yaml:3]: update

File: tests/migration/openldap_to_389ds.pm
Comment (coolgw) [reply tests/migration/openldap_to_389ds.pm:30]: I would like to see the screen instead of serial info:)

File: tests/migration/openldap_to_389ds.pm
Comment (coolgw) [reply tests/migration/openldap_to_389ds.pm:37]: Yes, this package will install clear cache tools for sssd

File: data/openldap_to_389ds/slapd.conf
Comment (coolgw) [reply data/openldap_to_389ds/slapd.conf:38]: No TLS currently, by plan

File: tests/migration/openldap_to_389ds.pm
Comment (coolgw) [reply tests/migration/openldap_to_389ds.pm:34]: just check version make sure latest build is ok when you do some debugging or regression run.

File: tests/migration/openldap_to_389ds.pm
Comment (coolgw) [reply tests/migration/openldap_to_389ds.pm:82]: That's trick part, sometimes the getent seems not stable, so i put some sleep here.

File: tests/migration/openldap_to_389ds.pm
Comment (foursixnine) [reply tests/migration/openldap_to_389ds.pm:30]: @coolgw how about switching to serial terminal before merging?

File: data/openldap_to_389ds/sssd.conf
Comment (coolgw) [reply data/openldap_to_389ds/sssd.conf:19]: Not enable security currently.

File: tests/migration/openldap_to_389ds.pm
Comment (foursixnine) [reply tests/migration/openldap_to_389ds.pm:82]: If that's the reason, express it better in the comment:

```
Wait an arbitrary ammount of time to give getent time be able to locate the users in the database
```

File: tests/migration/openldap_to_389ds.pm
Comment (foursixnine) [reply tests/migration/openldap_to_389ds.pm:34]: Then do it rather `zypper_call('info 389-ds');` in the end we'd get: https://openqa.suse.de/tests/5591428#step/firewalld/6

File: tests/migration/openldap_to_389ds.pm
Comment (rfan1) [reply tests/migration/openldap_to_389ds.pm:82]: Issue "sync" command may help there, in cases the database is not updated in short time. we can have a try :)

File: data/openldap_to_389ds/sssd.conf
Comment (rfan1) [reply data/openldap_to_389ds/sssd.conf:19]: Then we can remove this line safely

File: data/openldap_to_389ds/slapd.conf
Comment (rfan1) [reply data/openldap_to_389ds/slapd.conf:38]: Thanks, then I suggest removingt these lines

File: tests/migration/openldap_to_389ds.pm
Comment (coolgw) [reply tests/migration/openldap_to_389ds.pm]: I will create issue for this later, remove this line

**Comment by coolgw** [comment]
@rfan1 @foursixnine @b10n1k  update , could you help recheck it?

**Review by foursixnine — APPROVED** [review]

File: tests/migration/openldap_to_389ds.pm
Comment (foursixnine) [inline tests/migration/openldap_to_389ds.pm:32]:
```suggestion
    my $password = $testapi::password;
```

**Review by foursixnine — CHANGES_REQUESTED** [review]
Only this minor change in the password and we're good on my side

File: tests/migration/openldap_to_389ds.pm
Comment (rfan1) [inline tests/migration/openldap_to_389ds.pm:40]: You installed "sssd-tools" twice, any reason?

File: tests/migration/openldap_to_389ds.pm
Comment (rfan1) [inline tests/migration/openldap_to_389ds.pm:48]: I think we can remove this line, since it can work file now, however, depend on you

File: schedule/migration/openldap_to_389ds.yaml
Comment (okurz) [reply schedule/migration/openldap_to_389ds.yaml:3]: there is still a double space between "for" and "openldap"

Comment (okurz) [reply schedule/migration/openldap_to_389ds.yaml]: Also I think a description with "This is for" is not really helpful. Better describe what this scenario is actually doing

File: tests/migration/openldap_to_389ds.pm
Comment (okurz) [inline tests/migration/openldap_to_389ds.pm:16]: I don't understand why you write "Implement ... into openQA". This should be a summary of what the test module is doing, not a description of what your pull request is introducing

File: schedule/migration/openldap_to_389ds.yaml
Comment (okurz) [inline schedule/migration/openldap_to_389ds.yaml:3]: That should be rewritten not to say "This is" but rather what this scenario does and what it covers

File: tests/migration/openldap_to_389ds.pm
Comment (okurz) [inline tests/migration/openldap_to_389ds.pm:48]: I suggest to use `assert_script_run`. These calls should not fail, right? Same below

**Review by okurz — CHANGES_REQUESTED** [review]
What about a corresponding TW test? Because of [Factory First](https://opensource.suse.com/suse-open-source-policy) we should test against Tumbleweed by default (with exceptions for special packages, of course). Could you please check if it works with a current TW snapshot and then add to the main.pm there?

See https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/CONTRIBUTING.md#coding-style for more details

File: tests/migration/openldap_to_389ds.pm
Comment (okurz) [inline tests/migration/openldap_to_389ds.pm:43]: Why would Tumbleweed need to be handled different here?

Our tests should always consider openSUSE Tumbleweed as the default product without a version. That means any differing behaviour should have an explicit exclude rule for the "older products" and potentially an explanation why the behaviour should differ. You can try to negate the logic check.

See https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/CONTRIBUTING.md#coding-style for more details

**Review by okurz — CHANGES_REQUESTED** [review]

File: tests/migration/openldap_to_389ds.pm
Comment (coolgw) [reply tests/migration/openldap_to_389ds.pm:43]: I am debugging on this now, since i encounter no /etc/nsswitch.conf error when i running tumbleweed(So i need try add extra package in tumbleweed).  I can set this PR to WIP currently.  Will notify you once i am ready.

Comment (foursixnine) [reply tests/migration/openldap_to_389ds.pm:43]: > I am debugging on this now, since i encounter no /etc/nsswitch.conf error when i running tumbleweed(So i need try add extra package in tumbleweed). I can set this PR to WIP currently. Will notify you once i am ready.

I think that in TW you will find the system files in: /usr/etc/nsswitch.conf, https://bugzilla.opensuse.org/show_bug.cgi?id=1175238#c10

So you will need to create the file if doesn't exit, or copy it from usr/etc if we're on sle

Comment (coolgw) [reply tests/migration/openldap_to_389ds.pm:43]: @okurz @foursixnine  TW also pass now  https://openqa.opensuse.org/tests/1663905#
could you help have a check again?

**Review by foursixnine — APPROVED** [review]

**Comment by foursixnine** [comment]
I'll wait for @okurz's review by EBD before merging, in case he has an extra comment on this regard.

**Review by okurz — CHANGES_REQUESTED** [review]
The commit message could have a bit more details, e.g. what issue is fixed, why this needs to change, to which versions of which product it applies, link to a bug or a fate entry, the choices you made, etc. Also see https://commit.style/ or http://chris.beams.io/posts/git-commit/ as a helpful guide how to write good commit messages.

File: tests/migration/openldap_to_389ds.pm
Comment (okurz) [reply tests/migration/openldap_to_389ds.pm:43]: Still, treating Tumbleweed as something special is wrong here. As written in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/12075/files#r590125044 Tumbleweed just happens to have more recent versions of packages but again: Keep a version check, but not use the "always-rolling" Tumbleweed for a version check, use a negated version check, e.g.

```suggestion
    zypper_call 'in glibc' unless is_sle('<15-sp3') or is_leap('<15.3');
```

**Review by okurz — CHANGES_REQUESTED** [review]

**Comment by okurz** [comment]
@foursixnine unless my coffee is not strong enough I consider https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/12075#pullrequestreview-607186708 still unresolved :)

File: tests/migration/openldap_to_389ds.pm
Comment (coolgw) [reply tests/migration/openldap_to_389ds.pm]: Sorry forget remove this debug info. We do not need this anymore. 
also with latest verify run:
https://openqa.opensuse.org/t1668418

**Review by okurz — APPROVED** [review]

**Pull request merged**
